File: intact_agent_runner/files.py
from __future__ import annotations

from pathlib import Path


def read_text(file_path: str | Path) -> str:
    return Path(file_path).read_text(encoding="utf-8")


def write_text(file_path: str | Path, content: str) -> None:
    path = Path(file_path)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")


def list_markdown(root: str | Path) -> list[dict]:
    root_path = Path(root)
    if not root_path.exists():
        return []
    results: list[dict] = []
    for path in root_path.rglob("*.md"):
        if ".git" in path.parts or "node_modules" in path.parts:
            continue
        stat = path.stat()
        results.append({
            "absolute": str(path),
            "relative": str(path.relative_to(root_path)),
            "mtimeMs": stat.st_mtime * 1000,
        })
    return sorted(results, key=lambda item: item["relative"])


def latest_markdown(root: str | Path, limit: int = 5) -> list[dict]:
    files = list_markdown(root)
    return sorted(files, key=lambda item: item["mtimeMs"], reverse=True)[:limit]

File: src/files.js
import { mkdir, readFile, readdir, stat, writeFile } from "node:fs/promises";
import { existsSync } from "node:fs";
import path from "node:path";


export async function readText(filePath) {
  return readFile(filePath, "utf8");
}

export async function writeText(filePath, content) {
  await mkdir(path.dirname(filePath), { recursive: true });
  await writeFile(filePath, content, "utf8");
}

export async function listMarkdown(root) {
  if (!existsSync(root)) return [];
  const results = [];
  async function walk(current) {
    const entries = await readdir(current, { withFileTypes: true });
    for (const entry of entries) {
      const absolute = path.join(current, entry.name);
      if (entry.isDirectory()) {
        if (entry.name === ".git" || entry.name === "node_modules") continue;
        await walk(absolute);
      } else if (entry.isFile() && entry.name.endsWith(".md")) {
        const info = await stat(absolute);
        results.push({
          absolute,
          relative: path.relative(root, absolute),
          mtimeMs: info.mtimeMs,
        });
      }
    }
  }
  await walk(root);
  return results.sort((a, b) => a.relative.localeCompare(b.relative));
}

export async function latestMarkdown(root, limit = 5) {
  const files = await listMarkdown(root);
  return files.sort((a, b) => b.mtimeMs - a.mtimeMs).slice(0, limit);
}
